sites/funmooc/src/backend/base/cache.py

```diff
@@ -0,0 +1,111 @@
+"""
+  RedisCache with a fallback cache to prevent deny of service if Redis is downed
+  Freely inspired by django-cache-fallback
+  https://github.com/Kub-AT/django-cache-fallback/blob/master/cache_fallback/cache.py
+"""
+
+import logging
+
+from django.conf import settings
+from django.core import cache, signals
+from django.core.cache.backends.base import BaseCache
+
+from django_redis.cache import RedisCache
+
+DJANGO_REDIS_LOGGER = getattr(settings, "DJANGO_REDIS_LOGGER", __name__)
+
+logger = logging.getLogger(__name__)
+
+
+class RedisCacheWithFallback(BaseCache):
+    def __init__(self, server, params):
+        super().__init__(params)
+        options = params.get("OPTIONS", {})
+        self._redis_cache = RedisCache(server, params)
+        self._fallback_cache = self.initialize_fallback_cache(
+            options.get(
+                "FALLBACK_CLASS", "django.core.cache.backends.locmem.LocMemCache"
+            )
+        )
+        self.log = logging.getLogger(DJANGO_REDIS_LOGGER)
+
+    @staticmethod
+    def initialize_fallback_cache(backend, **kwargs):
+        fallback_cache = cache._create_cache(backend, **kwargs)
+        # Some caches -- python-memcached in particular -- need to do a cleanup at the
+        # end of a request cycle. If not implemented in a particular backend
+        # cache.close is a no-op. Not available in Django 1.5
+        if hasattr(fallback_cache, "close"):
+            signals.request_finished.connect(fallback_cache.close)
+        return fallback_cache
+
+    def _call_with_fallback(self, method, *args, **kwargs):
+        try:
+            return self._call_redis_cache(method, args, kwargs)
+        except Exception as e:
+            self.log.warning("[DEGRADED CACHE MODE] - Switch to fallback cache")
+            self.log.exception(e)
+            return self._call_fallback_cache(method, args, kwargs)
+
+    def _call_redis_cache(self, method, args, kwargs):
+        return getattr(self._redis_cache, method)(*args, **kwargs)
+
+    def _call_fallback_cache(self, method, args, kwargs):
+        return getattr(self._fallback_cache, method)(*args, **kwargs)
+
+    def get_backend_timeout(self, *args, **kwargs):
+        return self._call_with_fallback("get_backend_timeout", *args, **kwargs)
+
+    def make_key(self, *args, **kwargs):
+        return self._call_with_fallback("add", *args, **kwargs)
```

Comment:
should it be make_key?
```suggestion
        return self._call_with_fallback("make_key", *args, **kwargs)
```

Reply:
Copy pasta failed